### PR TITLE
feat: dynamic data in custom fields

### LINKED ISSUE #240: F11: Dynamic data in custom fields
We are adding dynamic data fields to custom forms,

The supported data types so far are:

- customers
- activities
- product variants

### PR DESCRIPTION
**Pull request by callmekungfu**
### Related Issue(s)

resolves #240 

### Quick overview of the change

- Added the ability to create dynamic data questions
- Added the ability to respond and render dynamic data questions
- Added basic documentations for custom forms

### CODE REVIEW

File: client/src/components/form-viewer/form-fields-factory.tsx [hunk 1]
```diff
@@ -104,3 +119,102 @@ export const MultiSelectFactory: FieldFactoryFunction = (
...
+      return;
+    }
+    setValue(id, data);
+  };
+
+  if (Array.isArray(options)) {
+    return (
+      <Alert severity="error">
+        There is an issue with this field. Please contact your administrator for
+        help
```
Comment (Wuon) [inline client/src/components/form-viewer/form-fields-factory.tsx:142]: missing period in this alert

File: server/src/models/forms/form.controller.ts [hunk 2]
```diff
@@ -108,7 +115,10 @@ export class FormController {
     @Param('id') id: string,
     @Body() { body, ...data }: UpdateFormDto,
   ) {
+    console.log(validateWithLatest);
```
Comment (Wuon) [inline server/src/models/forms/form.controller.ts:118]: wee woo console.log alert found

File: server/src/models/forms/form.controller.ts [hunk 3]
```diff
@@ -108,7 +115,10 @@ export class FormController {
...
+
     if (!validateWithLatest(body)) {
+      console.log(validateWithLatest.errors);
```
Comment (Wuon) [inline server/src/models/forms/form.controller.ts:121]: here too

**Review by Wuon — CHANGES_REQUESTED** [review]

File: client/src/components/form-builder/components/StaticMultipleChoiceBuilder.tsx [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+
+export const StaticMultipleChoiceBuilder = ({
+  context,
+  sectionIndex,
+  index,
+  control,
+}: StaticBuilderProps) => {
+  const optionData = useWatch({
+    control,
+    name: `${context}sections.${sectionIndex}.fields.${index}.options`,
```
[the remark this answers is not in the document]
Comment (callmekungfu) [reply client/src/components/form-builder/components/StaticMultipleChoiceBuilder.tsx:23]: You mean make the name generation a function? I do that if the name gets used repeatedly, in this case it doesn't so we keep it inline

File: client/src/components/form-builder/form-field.tsx [hunk 5]
```diff
@@ -41,23 +39,18 @@ export const FormField = ({
...
+
+  const isStaticOptionQuestionType =
     type === 'multipleChoice' || type === 'multiSelect';
-  const { append, remove, fields } = useFieldArray({
-    control,
-    name: `${context}sections.${sectionIndex}.fields.${index}.options`,
-  });
-  const isFieldOptionsAnArray = useMemo(() => Array.isArray(fields), [fields]);
+
+  const isDynamicOptionQuestionType = type === 'dataSourceSelect';
```
[the remark this answers is not in the document]
Comment (callmekungfu) [reply client/src/components/form-builder/form-field.tsx:52]: Good point, let me see if I can enumerate them in c-form and pass that around

File: client/src/components/form-builder/components/StaticMultipleChoiceBuilder.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (Wuon) [reply client/src/components/form-builder/components/StaticMultipleChoiceBuilder.tsx:23]: why is everything else `.` notation except for first part?

eg:
`${context}sections` vs `${context}.sections`

**Review by Wuon — APPROVED** [review]

**Pull request merged**